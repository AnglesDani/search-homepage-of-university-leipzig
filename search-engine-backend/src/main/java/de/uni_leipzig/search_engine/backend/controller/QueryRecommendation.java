package de.uni_leipzig.search_engine.backend.controller;

import java.util.List;

import lombok.Data;

@Data
public class QueryRecommendation
{
	private final List<String> recommendations;
}
