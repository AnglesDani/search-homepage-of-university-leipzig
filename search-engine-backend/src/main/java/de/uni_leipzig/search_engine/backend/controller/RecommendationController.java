package de.uni_leipzig.search_engine.backend.controller;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.lucene.analysis.core.SimpleAnalyzer;
import org.apache.lucene.search.spell.Dictionary;
import org.apache.lucene.search.suggest.Lookup.LookupResult;
import org.apache.lucene.search.suggest.analyzing.FuzzySuggester;
import org.apache.lucene.store.RAMDirectory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

import lombok.SneakyThrows;

@Controller
public class RecommendationController
{
	private final FuzzySuggester fuzzySuggester = new FuzzySuggester(new RAMDirectory(), "sadsa", new SimpleAnalyzer());

	@Autowired
	@SneakyThrows
	public RecommendationController(Dictionary dict)
	{
		fuzzySuggester.build(dict);
	}
	
	@SneakyThrows
	@RequestMapping(method=RequestMethod.GET, path="queryRecommendation")
	public Object recommendForUserQuery(@RequestParam String query)
	{
		List<LookupResult> ret = fuzzySuggester.lookup(query, Boolean.FALSE, 10);
		
		return new QueryRecommendation(ret.stream()
				.map(i -> String.valueOf(i.key))
				.collect(Collectors.toList()));
	}
}
