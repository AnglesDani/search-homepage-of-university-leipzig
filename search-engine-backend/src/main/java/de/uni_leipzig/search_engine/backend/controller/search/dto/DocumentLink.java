package de.uni_leipzig.search_engine.backend.controller.search.dto;

import org.springframework.hateoas.Link;
import org.springframework.hateoas.mvc.ControllerLinkBuilder;

import de.uni_leipzig.search_engine.backend.controller.redirect.RedirectController;
import groovy.transform.EqualsAndHashCode;
import lombok.Data;

/**
 * A {@link Link} which contains a
 * {@link Link#getHref() tracking link} and a {@link DocumentLink#displayLink display link}.
 * 
 * @author Maik Fr√∂be
 *
 */
@Data
@EqualsAndHashCode(callSuper=true)
class DocumentLink extends Link
{
	private static final long serialVersionUID = 1L;

	private String displayLink;
	
	DocumentLink(Integer docID, Integer duplicate, String displayLink)
	{
		super(createTargetLink(docID, duplicate).getHref(), duplicate < 0 ? Link.REL_SELF : "duplicate");
		this.displayLink = displayLink;
	}
	
	private static Link createTargetLink(Integer docID, Integer duplicate)
	{
		return ControllerLinkBuilder.linkTo(
			ControllerLinkBuilder.methodOn(RedirectController.class).redirect(docID, duplicate))
			.withRel(Link.REL_SELF);
	}
}
