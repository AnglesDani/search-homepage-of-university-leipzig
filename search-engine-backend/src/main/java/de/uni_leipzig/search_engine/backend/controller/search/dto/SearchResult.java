package de.uni_leipzig.search_engine.backend.controller.search.dto;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Triple;
import org.apache.lucene.document.Document;
import org.springframework.hateoas.Link;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@Data
@NoArgsConstructor
@Accessors(chain=true)
public class SearchResult
{
	public static final String INDEX_FIELD_TITLE = "title";
	
	public static final String INDEX_FIELD_CONTENT = "content";
	
	public static final String INDEX_FIELD_LINK = "url";
	
	public static final String INDEX_FIELD_ANCHOR = "anchor";
	
	public static final String INDEX_FIELD_LINKS_TO_DUPLICATES = "duplicatesUrl";
	
	private String title;
	
	private String snippet;
	
	private Link targetUrl;
	
	private List<Link> linksToDuplicates;
	
	public SearchResult(Triple<Document, Integer, String> result)
	{
		setSnippet(result.getRight());
		setTargetUrl(new DocumentLink(result.getMiddle(), -1, result.getLeft().get(INDEX_FIELD_LINK)));
		setLinksToDuplicates(createLinksToDuplicates(result.getMiddle(), result.getLeft()));
		setTitle(result.getLeft().get(INDEX_FIELD_TITLE));
		
		
		if(getTitle() == null || getTitle().isEmpty())
		{
			setTitle(StringUtils.abbreviate(result.getLeft().get(INDEX_FIELD_CONTENT), 40));
		}
	}

	
	private static List<Link> createLinksToDuplicates(Integer docID, Document doc)
	{
		List<Link> ret = new ArrayList<>();
		String[] duplicateUrls = null;
		
		if(doc != null && (duplicateUrls = doc.getValues(INDEX_FIELD_LINKS_TO_DUPLICATES)) != null
				&& duplicateUrls.length > 0)
		{
			for(int i=0; i< duplicateUrls.length; i++)
			{
				ret.add(new DocumentLink(docID, i, duplicateUrls[i]));
			}
		}
		
		return ret;
	}
}
