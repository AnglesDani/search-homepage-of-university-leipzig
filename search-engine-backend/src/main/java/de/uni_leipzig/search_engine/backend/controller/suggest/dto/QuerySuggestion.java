package de.uni_leipzig.search_engine.backend.controller.suggest.dto;

import java.util.List;

import lombok.Data;

@Data
public class QuerySuggestion
{
	private final List<String> recommendations;
}
