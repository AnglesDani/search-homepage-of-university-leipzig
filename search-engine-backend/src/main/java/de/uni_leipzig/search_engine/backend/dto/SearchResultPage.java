package de.uni_leipzig.search_engine.backend.dto;

import java.util.List;

import org.springframework.hateoas.Link;

import lombok.Data;
import lombok.experimental.Accessors;

@Data
@Accessors(chain=true)
public class SearchResultPage
{
	private String originalQuery;

	private int page;
	
	private List<SearchResult> resultsOnPage;
	
	private long totalHits;
	
	private Link nextPage;
	
	private Link previousPage;
	
	private Link firstPageLink;
	
	private Link lastPageLink;
	
	private List<Link> namedPageLinksBefore;
	
	private List<Link> namedPageLinksAfter;
}
