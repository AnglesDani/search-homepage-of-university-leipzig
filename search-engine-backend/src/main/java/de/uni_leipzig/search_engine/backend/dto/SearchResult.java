package de.uni_leipzig.search_engine.backend.dto;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.lang3.tuple.Triple;
import org.apache.lucene.document.Document;
import org.springframework.hateoas.Link;
import org.springframework.hateoas.mvc.ControllerLinkBuilder;

import de.uni_leipzig.search_engine.backend.controller.RedirectController;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain=true)
public class SearchResult
{
	public static final String INDEX_FIELD_TITLE = "title";
	
	public static final String INDEX_FIELD_CONTENT = "content";
	
	public static final String INDEX_FIELD_LINK = "url";
	
	public static final String INDEX_FIELD_ANCHOR = "anchor";
	
	public static final String INDEX_FIELD_LINKS_TO_DUPLICATES = "duplicatesUrl";
	
	private String title;
	
	private String snippet;
	
	private Link targetUrl;
	
	private List<Link> linksToDuplicates;
	
	//FIXME remove this: only fast and ugly
	private String duplicateContent;
	
	public SearchResult(Triple<Document, Integer, String> result)
	{
		setTitle(result.getLeft().get(INDEX_FIELD_TITLE));
		setSnippet(result.getRight());
		setTargetUrl(createTargetLink(result.getMiddle()));
		setLinksToDuplicates(createLinksToDuplicates(result.getMiddle(), result.getLeft()));
		
		if(!getLinksToDuplicates().isEmpty())
		{
			duplicateContent = "<ul>"+ getLinksToDuplicates().stream()
					.map(l -> "<li><a href=\""+ l.getHref() +"\">"+ l.getRel() +"</a></li>")
					.collect(Collectors.joining(" "))
		
					+"</ul>";
			System.out.println(duplicateContent);
		}
	}
	
	private static Link createTargetLink(Integer docID)
	{
		return createTargetLink(docID, -1);
	}
	
	private static Link createTargetLink(Integer docID, Integer duplicate)
	{
		return ControllerLinkBuilder.linkTo(
			ControllerLinkBuilder.methodOn(RedirectController.class).redirect(docID, duplicate))
			.withRel("targetUrl");
	}
	
	private static List<Link> createLinksToDuplicates(Integer docID, Document doc)
	{
		List<Link> ret = new ArrayList<>();
		String[] duplicateUrls = null;
		
		if(doc != null && (duplicateUrls = doc.getValues(INDEX_FIELD_LINKS_TO_DUPLICATES)) != null
				&& duplicateUrls.length > 0)
		{
			for(int i=0; i< duplicateUrls.length; i++)
			{
				Link link = createTargetLink(docID, i);
				link = link.withRel(duplicateUrls[i]);
				
				ret.add(link);
			}
		}
		
		return ret;
	}
}
