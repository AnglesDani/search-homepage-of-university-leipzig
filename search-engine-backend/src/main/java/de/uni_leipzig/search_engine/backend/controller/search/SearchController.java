package de.uni_leipzig.search_engine.backend.controller.search;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import org.apache.commons.lang3.tuple.Pair;
import org.apache.commons.lang3.tuple.Triple;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.TopDocs;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.hateoas.Link;
import org.springframework.hateoas.LinkBuilder;
import org.springframework.hateoas.mvc.ControllerLinkBuilder;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

import de.uni_leipzig.search_engine.backend.controller.search.dto.SearchResult;
import de.uni_leipzig.search_engine.backend.controller.search.dto.SearchResultPage;
import de.uni_leipzig.search_engine.backend.lucene.HighlightComponent;
import de.uni_leipzig.search_engine.backend.lucene.SearcherComponent;
import lombok.SneakyThrows;

@Controller
public class SearchController
{
	@Autowired
	private SearcherComponent searcherComponent;
	
	@Autowired
	private HighlightComponent highlightComponent;

	@RequestMapping(method=RequestMethod.GET, path="/api/v1")
	public Object search(@RequestParam(defaultValue="") String q,
			@RequestParam(defaultValue="1") Integer p)
	{
		SearchResultPage ret = new SearchResultPage();
		ret.setOriginalQuery(q);
		p = Math.max(p == null? 1: p, 1);
		
		if("".equals(q) || q == null)
		{
			return ret;
		}
		
		int topNForSearchResult = Math.max(SearcherComponent.HITS_PER_PAGE, p * SearcherComponent.HITS_PER_PAGE);
		
		TopDocs queryTopDocs = searcherComponent.search(q, topNForSearchResult);
		Pair<Integer, List<ScoreDoc>> lastPageNumberAndContent = determineLastPageNumberAndContent(queryTopDocs);
		ret.setTotalHits(queryTopDocs.totalHits);
		ret.setResultsOnPage(mapScoreDocsToSearchResults(lastPageNumberAndContent.getRight(), q));
		ret.setPage(lastPageNumberAndContent.getLeft());
		injectPaginationLinks(ret);
		
		return ret;
	}
	
	public static Pair<Integer, List<ScoreDoc>> determineLastPageNumberAndContent(TopDocs topDocs)
	{
		if(topDocs == null || topDocs.scoreDocs == null || topDocs.scoreDocs.length == 0)
		{
			return Pair.of(1, new ArrayList<>());
		}
		
		final int page = ((topDocs.scoreDocs.length-1)/SearcherComponent.HITS_PER_PAGE);
		List<ScoreDoc> ret = new ArrayList<>();
		
		for(int i=page*SearcherComponent.HITS_PER_PAGE; i< topDocs.scoreDocs.length; i++)
		{
			ret.add(topDocs.scoreDocs[i]);
		}
		
		return Pair.of(page+1, ret);
	}
	
	private static LinkBuilder searchLink(String query, Integer page)
	{
		return ControllerLinkBuilder.linkTo(ControllerLinkBuilder.methodOn(SearchController.class).search(query, page));
	}
	
	@SneakyThrows
	private List<SearchResult> mapScoreDocsToSearchResults(List<ScoreDoc> scoreDocs, String query)
	{
		Query parsedQuery = highlightComponent.parsQueryForHiglights(query);
		
		return scoreDocs.stream()
			.map(topDoc -> Triple.of(searcherComponent.doc(topDoc.doc), topDoc.doc, 
					highlightComponent.buildHiglightForDocument(topDoc, parsedQuery)))
			.map(SearchResult::new)
			.collect(Collectors.toList());
	}

	public static void injectPaginationLinks(SearchResultPage searchResultPage)
	{
		if(searchResultPage == null)
		{
			return;
		}
		
		int currentPage = Math.max(searchResultPage.getPage(), 1);
		int maxPage = (((int)searchResultPage.getTotalHits()-1)/SearcherComponent.HITS_PER_PAGE) +1;
		
		if(currentPage > 1)
		{
			searchResultPage.setPreviousPage(searchLink(searchResultPage.getOriginalQuery(), currentPage -1).withRel("prev"));
		}
		
		if(currentPage < maxPage)
		{
			searchResultPage.setNextPage(searchLink(searchResultPage.getOriginalQuery(), currentPage+1).withRel("next"));
		}
		
		searchResultPage.setNamedPageLinksAfter(namedPaginationLinksInRange(currentPage+1,
			Math.min(currentPage+4, maxPage+1), searchResultPage));
		
		searchResultPage.setNamedPageLinksBefore(namedPaginationLinksInRange(Math.max(currentPage-3, 1),
				currentPage, searchResultPage));
		
		Link firstPageLink = searchLink(searchResultPage.getOriginalQuery(), 1).withRel("1");
		if(currentPage > 1 && !searchResultPage.getNamedPageLinksBefore().contains(firstPageLink))
		{
			searchResultPage.setFirstPageLink(firstPageLink);
		}
		
		Link lastPageLink = searchLink(searchResultPage.getOriginalQuery(), maxPage).withRel(String.valueOf(maxPage));
		if(currentPage < (maxPage-1) && !searchResultPage.getNamedPageLinksAfter().contains(lastPageLink))
		{
			searchResultPage.setLastPageLink(lastPageLink);
		}
	}
	
	private static List<Link> namedPaginationLinksInRange(int startInclusive, int endExclusive, SearchResultPage searchResultPage)
	{
		return IntStream.range(startInclusive, endExclusive)
			.mapToObj(i -> searchLink(searchResultPage.getOriginalQuery(), i).withRel(String.valueOf(i)))
			.collect(Collectors.toList());
	}
}
