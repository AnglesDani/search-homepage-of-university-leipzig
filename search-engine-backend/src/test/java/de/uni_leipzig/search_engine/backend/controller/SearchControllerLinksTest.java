package de.uni_leipzig.search_engine.backend.controller;

import java.util.Arrays;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import de.uni_leipzig.search_engine.backend.controller.SearchController;
import de.uni_leipzig.search_engine.backend.dto.SearchResultPage;

public class SearchControllerLinksTest
{
	@Before
	public void setup()
	{
		SearchControllerMockUtils.setupHttpServlet();
	}
	
	@Test
	public void testNullAsInput()
	{
		SearchController.injectPaginationLinks(null);
	}
	
	@Test
	public void testEmptyPageAsInput()
	{
		SearchResultPage expected = new SearchResultPage();
		expected.setNamedPageLinksBefore(Arrays.asList());
		expected.setNamedPageLinksAfter(Arrays.asList());
		
		SearchResultPage resultPageToVerify = new SearchResultPage();
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Assert.assertEquals(expected, resultPageToVerify);
	}
	
	@Test
	public void checkEmptyResultAsInput()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(1);
		resultPageToVerify.setTotalHits(0);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatNextPageLinkIsCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(1);
		resultPageToVerify.setTotalHits(11);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatPreviousPageLinkIsCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(2);
		resultPageToVerify.setTotalHits(11);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatNextAndPreviousPageLinksAreCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(2);
		resultPageToVerify.setTotalHits(21);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatNextPageLinkAndNamedPageLinksAfterAreCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(1);
		resultPageToVerify.setTotalHits(61);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatPreviousPageLinkAndNamedPageLinksBeforeAreCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(7);
		resultPageToVerify.setTotalHits(61);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatNoFirstAndLastPageLinkIsCreatedIfTheyAreAlreadyContainedInNamedLinks()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(4);
		resultPageToVerify.setTotalHits(70);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
	
	@Test
	public void checkThatFirstAndLastPageLinkAndBeforeAndAfterLinksAreCreated()
	{
		SearchResultPage resultPageToVerify = new SearchResultPage();
		resultPageToVerify.setPage(5);
		resultPageToVerify.setTotalHits(100);
		SearchController.injectPaginationLinks(resultPageToVerify);
		
		Approvals.verify(resultPageToVerify);
	}
}
