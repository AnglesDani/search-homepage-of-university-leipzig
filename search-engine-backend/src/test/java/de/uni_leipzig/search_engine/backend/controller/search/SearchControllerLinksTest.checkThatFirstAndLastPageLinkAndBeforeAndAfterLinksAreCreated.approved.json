{
  "page": 5,
  "totalHits": 100,
  "nextPage": {
    "rel": "next",
    "href": "http://www.example.com/api/v1/search?p\u003d6",
    "template": {
      "variables": {
        "variables": []
      },
      "baseUri": "http://www.example.com/api/v1/search?p\u003d6"
    }
  },
  "previousPage": {
    "rel": "prev",
    "href": "http://www.example.com/api/v1/search?p\u003d4",
    "template": {
      "variables": {
        "variables": []
      },
      "baseUri": "http://www.example.com/api/v1/search?p\u003d4"
    }
  },
  "firstPageLink": {
    "rel": "1",
    "href": "http://www.example.com/api/v1/search?p\u003d1",
    "template": {
      "variables": {
        "variables": []
      },
      "baseUri": "http://www.example.com/api/v1/search?p\u003d1"
    }
  },
  "lastPageLink": {
    "rel": "10",
    "href": "http://www.example.com/api/v1/search?p\u003d10",
    "template": {
      "variables": {
        "variables": []
      },
      "baseUri": "http://www.example.com/api/v1/search?p\u003d10"
    }
  },
  "namedPageLinksBefore": [
    {
      "rel": "2",
      "href": "http://www.example.com/api/v1/search?p\u003d2",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d2"
      }
    },
    {
      "rel": "3",
      "href": "http://www.example.com/api/v1/search?p\u003d3",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d3"
      }
    },
    {
      "rel": "4",
      "href": "http://www.example.com/api/v1/search?p\u003d4",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d4"
      }
    }
  ],
  "namedPageLinksAfter": [
    {
      "rel": "6",
      "href": "http://www.example.com/api/v1/search?p\u003d6",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d6"
      }
    },
    {
      "rel": "7",
      "href": "http://www.example.com/api/v1/search?p\u003d7",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d7"
      }
    },
    {
      "rel": "8",
      "href": "http://www.example.com/api/v1/search?p\u003d8",
      "template": {
        "variables": {
          "variables": []
        },
        "baseUri": "http://www.example.com/api/v1/search?p\u003d8"
      }
    }
  ]
}